### PR TITLE
Allow consumers to set always_include_developer_search_paths in swift_library

### PR DESCRIPTION
**Pull request by thiagohmcruz**
https://github.com/bazelbuild/rules_swift/pull/1162

### CODE REVIEW

**Review by AngelaGuardia — APPROVED** [review]

File: rules/library.bzl
Comment (mattrobmattrob) [inline rules/library.bzl:988]: I suppose we only need the rules_swift bump because we're applying this directly vs. in the swift_kwargs below.

Is it easy to not require the rules_swift bump? I don't really feel strongly either way just wanted to ask.

**Review by mattrobmattrob — APPROVED** [review]

File: rules/library.bzl
Comment (thiagohmcruz) [reply rules/library.bzl:988]: That's a good callout, aligned with feedback I got from @luispadron too. I guess one looking to use the flag can simply use kwargs indeed. 

For those looking to use the flag It would simplify things if `rules_ios` was on this `rules_swift` version though. So i guess I can repurpose this PR to simply bump `rules_swift` if folks don't have any concerns with that.

Comment (thiagohmcruz) [reply rules/library.bzl:988]: @mattrobmattrob Q: What would be the issue with the `rules_swift` bump here specifically? I could make the changes in a way that for us everything works. But isn't this a good opportunity to add support for this flag for all consumers?

Comment (mattrobmattrob) [reply rules/library.bzl:988]: Mostly just that this requires a rules_swift bump for consumers. Not necessarily a bad thing though.

**Comment by thiagohmcruz** [comment]
Discussed offline and we agreed that instead of bumping `rules_swift` in `rules_ios` it provides a better DX for both consumers and maintainers to simply allow the flag `always_include_developer_search_paths`to be set in kwargs and propagated to the `swift_library`. Letting consumers decide if they want to bump to a newer `rules_swift` version or not to start setting that flag in their BUILD files.

**Review by luispadron — APPROVED** [review]
Thanks for iterating on this!

**Comment by mattrobmattrob** [comment]
Shouldn't the flag be `always_include_developer_search_paths` instead of `swift_ always_include_developer_search_paths`?

File: rules/library.bzl
Comment (mattrobmattrob) [inline rules/library.bzl:965]: Want to do the if check first before mutating the `swift_kwargs` to minimize the penalty to all callers?

**Review by mattrobmattrob — APPROVED** [review]
I do wonder if we should come up with a general approach for passing rule attributes through to this point vs. a one off flag. Doesn't seem like we have a pattern for it at this point though.

**Comment by thiagohmcruz** [comment]
> Shouldn't the flag be `always_include_developer_search_paths` instead of `swift_ always_include_developer_search_paths`?

@mattrobmattrob From a mixed frameworks perspective I think we should explicitly convey that this is a Swift thing and not Obj-C. Mostly thinking of a rules_ios user that is not aware of all the underlying impl details of the rules and want it to "just work". In that scenario the names used in the API are the main way to interact with the rule set and IMO we should be very explicit here. 

Open to change it though in case I am missing something here!

File: rules/library.bzl
Comment (thiagohmcruz) [reply rules/library.bzl:965]: Makes sense, we should do the same in the `plugins` line above in that case

**Comment by thiagohmcruz** [comment]
> I do wonder if we should come up with a general approach for passing rule attributes through to this point vs. a one off flag. Doesn't seem like we have a pattern for it at this point though.

@mattrobmattrob +1, I would still prefer to land this change to unblock ongoing work but I'm open to contribute to an issue/proposal on that. Def something I thought about in the past too while working on the rules.

**Comment by mattrobmattrob** [comment]
> > Shouldn't the flag be `always_include_developer_search_paths` instead of `swift_ always_include_developer_search_paths`?
> 
> 
> 
> @mattrobmattrob From a mixed frameworks perspective I think we should explicitly convey that this is a Swift thing and not Obj-C. Mostly thinking of a rules_ios user that is not aware of all the underlying impl details of the rules and want it to "just work". In that scenario the names used in the API are the main way to interact with the rule set and IMO we should be very explicit here. 
> 
> 
> 
> Open to change it though in case I am missing something here!


Deleted my comment but wasn't fast enough. I thought you were passing it through to the swift_library with the prefix but realized I was wrong.

File: rules/library.bzl
Comment (luispadron) [inline rules/library.bzl:582]: I agree this should be 
```suggestion
    swift_always_include_developer_search_paths = kwargs.pop("always_include_developer_search_paths", None)
```

we didnt make that distinction for `plugins`

File: rules/library.bzl
Comment (thiagohmcruz) [reply rules/library.bzl]: https://github.com/bazel-ios/rules_ios/pull/852/commits/71de457039f733c1b871f15e3d8f5842939647c1

File: rules/library.bzl
Comment (thiagohmcruz) [reply rules/library.bzl:582]: See [this](https://github.com/bazel-ios/rules_ios/pull/852#issuecomment-2018890660)

Comment (luispadron) [reply rules/library.bzl:582]: Essentially a user would have to look at the code here to understand they should use `swift_always_include_developer_search_paths` vs. the actual flag `always_include_developer_search_paths`

Comment (thiagohmcruz) [reply rules/library.bzl:582]: Ok, people will try to use flags by the original names. I'm convinced, cc @mattrobmattrob haha

Comment (thiagohmcruz) [reply rules/library.bzl]: https://github.com/bazel-ios/rules_ios/pull/852/commits/4a57fd23f7b82dfb8d26c33599dacc866a6f6c0e

**Pull request merged**
